### ISSUE TITLE
ULONGLONG to uintptr_t conversion causes DEBUG warning

### ISSUE THREAD

**Issue by Brett208** [OutpostUniverse/NetHelper#5]
The following warning is produced only when compiling in debug mode. The warning does not appear to exist in ReleaseMinSize builds.

Warning	C4244	'=': conversion from 'ULONGLONG' to 'uintptr_t', possible loss of data	NetHelper	\nethelper\src\patcher.cpp	570

Produced by the line ->ImageBase below.

``` c++
uintptr_t preferredAddress;

 . . .

    else if (optionalHeader->Magic == IMAGE_NT_OPTIONAL_HDR64_MAGIC) {
      preferredAddress = reinterpret_cast<IMAGE_OPTIONAL_HEADER64*>(optionalHeader)
                         ->ImageBase;
    }

. . .

modulePrefAddr[module] = preferredAddress;
```

**Comment by DanRStevens** [comment]
For reference:
[IMAGE_OPTIONAL_HEADER](https://docs.microsoft.com/en-us/windows/desktop/api/winnt/ns-winnt-_image_optional_header)
[IMAGE_OPTIONAL_HEADER64](https://docs.microsoft.com/en-us/windows/desktop/api/winnt/ns-winnt-_image_optional_header64)

The `ImageBase` value is effectively a pointer. The 32-bit version of this struct declares this field as a DWORD. The 64-bit version of this struct declares this field as a ULONGLONG.

This structure represents an optional part of the PE header of executable files on disk. The size of the fields is a property of the file itself (as in, which platform it was compiled for). It is not related to the pointer size of the current platform, which is storing the file.

The `uintptr_t` type represents an integer which is large enough to hold a pointer on the target platform.

A 32-bit build of *this* program will not have a large enough field to process the header of a 64-bit executable file (processed like a data file). As the code is using the 64-bit version of the struct specifically in the cast, it may be more appropriate to use a `ULONGLONG` here.

**Comment by Brett208** [comment]
Thanks for commenting on this. It makes a little more sense to me now. Basically, the code is checking whether the file header is 32 or 64 bit and then pulling the ImageBase from the appropriate IMAGE_OPTIONAL_HEADER version.

To make sure we can fit all possible values of a IMAGE_OPTIONAL_HEADER64, we should use a ULONGLONG.

I'm not ready to update the variable in this case though, as I don't understand how updating the type `preferredAddress' will affect its use later in the function.

In particular:
``` C++
  modulePrefAddr[module] = preferredAddress;
  }
  else {
    preferredAddress = it->second;
  }

  return reinterpret_cast<void*>(reinterpret_cast<uintptr_t>(pointer) -
                                 preferredAddress +
                                 reinterpret_cast<uintptr_t>(module));
```

modulePrefAddr is defined as:

``` C++
std::unordered_map<HMODULE, uintptr_t> modulePrefAddr
```

An HMODULE is defined as:

``` C++
DECLARE_HANDLE(HINSTANCE);
typedef HINSTANCE HMODULE;
```

I'm not really sure what is happening in the macro DECLARE_HANDLE, but I'm worried that HINSTANCE is only DWORD in size (not sure though). 

No idea how to handle the function return or the iterator modulePrefAddr if it is a ULONGLONG???

---

What if we just band aid by rejecting imageBases that are too large?

``` C++
#include <cstdint>

else if (optionalHeader->Magic == IMAGE_NT_OPTIONAL_HDR64_MAGIC) {
	ULONGLONG preferredAddress64 = reinterpret_cast<IMAGE_OPTIONAL_HEADER64*>(optionalHeader)->ImageBase;

	if (preferredAddress64 > MAXUINT32) {
		return nullptr; // Image base must fit in a DWORD
	}

	preferredAddress = static_cast<uintptr_t>(preferredAddress64);
}
```

Thanks,
Brett

**Comment by DanRStevens** [comment]
For reference [Windows Data Types](https://docs.microsoft.com/en-us/windows/desktop/winprog/windows-data-types)

From the above, you can trace back the following:
```cpp
typedef void* PVOID;
typedef PVOID HANDLE;
typedef HANDLE HINSTANCE;
```

A `HANDLE` is typically an opaque pointer, and as such, tends to be pointer sized, as can be seen from the typedefs above.

Also from the above, an `HMODULE` is effectively the same as an `HINSTANCE`, though that was not the case way back in 16-bit Windows.

----

If the bit size of the target module is the same as the bit size of the current module, which should be the case for in-process patching, then you should be able to declare the variable as a `HINSTANCE` (or `HMODULE`, or `HANDLE`).

**Comment by Brett208** [comment]
Can I assume that the size of PVOID is 32 bits on win32 and 64 bits on x64 compiled code?

If so, then I'm guessing we cannot use a HANDLE to represent the ImageBase of a file saved in a x64 IMAGE_FILE_HEADER format if the NetHelper DLL is compiled using win32.

-Brett

**Comment by DanRStevens** [comment]
It seems I missed out on an important point in my first reply.

Yes, a PVOID will depend on the pointer size, which is the main difference between 32-bit and 64-bit code.

If the patcher is used for in-process patching, then both the target module's bit-size, and the current module's bit-size should be the same. All modules loaded into the same process will be required to have the same bit-size. This is because all modules within a process must share the same address space, and the size of the address space corresponds with the pointer size.

You should be able to rely on compile time bit-size settings in this case. Hence a `HANDLE` type variable would be appropriate.

This would be different for out of process handling of an executable module, as an external data file. We are not doing that here.

**Comment by DanRStevens** [comment]
Ok, I took a closer look at the code. I believe the original code is actually good as written, with no chance for the warning to trigger bad behaviour. The reason the code is good, is that it does in-process patching, and so the loaded module must have a bit-size identical to the current module. Hence a `uintptr_t` type will be large enough to hold the value.

The code should of course have `#include <cstdint>` to access the `uintptr_t` type, as you've added.
Reference: [Fixed width integer types](https://en.cppreference.com/w/cpp/types/integer)

----

The reason for the warning, is the compiler can't figure out the 64-bit branch can never be taken by code compiled in 32-bit mode. The `module` parameter is an opaque `HANDLE` type, however, it is casted to its raw pointer form in that method, and used to access in-memory data structures describing the currently loaded module. For that cast to be valid, and point to module header data, the module must already be loaded into memory. That means its bit-size must be equal to the current module, otherwise it would have failed to load. (Based on current Win32 versus Win64 differences). Hence the module size checks could potentially be done as compile time `#ifdef`s rather than a runtime `if`-`else if` statement. The code in question being:
```cpp
if (optionalHeader->Magic == IMAGE_NT_OPTIONAL_HDR32_MAGIC) {
  ...
}
else if (optionalHeader->Magic == IMAGE_NT_OPTIONAL_HDR64_MAGIC) {
  ...
}
```

----

I see 2 possible solutions. The easy familiar way is to add a cast to `uintptr_t` type around the `ImageBase` value. An alternative is to use `#ifdef` checks around that code to replace the `if`-`else if`, so only the currently relevant branch of the code is compiled into the program.

Option 1: Silence the warning on Win32 within the dead Win64 code branch with a cast:
```cpp
preferredAddress = reinterpret_cast<uintptr_t>(reinterpret_cast<IMAGE_OPTIONAL_HEADER64*>(optionalHeader)->ImageBase);
```

Option 2: Conditionally define code to match the current platform:
```cpp
#ifdef _WIN64
  preferredAddress = reinterpret_cast<IMAGE_OPTIONAL_HEADER64*>(optionalHeader)->ImageBase;
#elif
  preferredAddress = optionalHeader->ImageBase;
#endif
```

They each have their downsides should a new platform be developed, or mixing of modules someday be allowed. I don't think we need to worry about that happening though.

**Comment by Brett208** [comment]
I think the preprocessor directive makes more sense to me. To me, it expresses more that the compilation matters than using a cast.

I may attempt to tackle integrating in the near future. If I start working on it, will assign myself to the issue to prevent duplicate work.

**Comment by DanRStevens** [comment]
I think I agree with your assessment. Thanks for the feedback. I went ahead and made the change.

While making the change, I noticed the original `if`-`else if` block had a sanity checking `else` block, which I preserved. I think that was actually one of my concerns about the change, was potentially losing such a sanity check. As such, I feel even more strongly this is the correct decision here.

**Comment by Arklon1** [comment]
Maybe I should actually start reading these...

Hooman's analysis is correct, the warning is benign because this logic will realistically only be hit in 64-bit builds, in which case uintptr_t is defined as uint64 instead of uint32 so there would be no warning anymore.
The fact that there's any 64-bit support at all is a bit silly, but I had implemented it anyway because I'm a completionist. Well, except for the assembly patching functions where I decided to have no principles and be lazy instead.
